### PR TITLE
Add Concept Guide

### PR DESCRIPTION
**Pull request by JohnSundell**
This guide aims to provide an easy-to-navigate reference of all the concepts that the Hub Framework uses. Contains lightweight explanations of things like “View models” or “Components”.

### CODE REVIEW

**Comment by JohnSundell** [comment]
@spotify/objc-dev The recommended way to review this is by reading the [rendered markdown](https://github.com/JohnSundell/HubFramework/blob/77680f58d5e95411af2fcfe77a70735dbc005409/documentation/Concept%20guide.md).

File: documentation/Concept guide.md [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+- The [Content programming guide](https://spotify.github.io/HubFramework/content-programming-guide.html).
+
+## Factory
+
+An object that is responsible for creating implementations of a certain protocol. The [factory pattern](https://en.wikipedia.org/wiki/Factory_method_pattern) is used extensively throughout the Hub Framework to avoid shared state. By always creating unique object instances for each context - no state is ever shared. While this requires a bit more code to be written (you have to write both an implementation **and a factory**, for example for [components](#component)), it makes the framework a lot more predictable and less error prone.
+
+For more information see:
+
+- ["Factory method pattern" on Wikipedia](https://en.wikipedia.org/wiki/Factory_method_pattern).
+- The  `HUBComponentFactory` protocol.
```
Comment (cerihughes) [inline documentation/Concept guide.md]: Nitpick: Extra space after `The`

File: documentation/Concept guide.md [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+- The `HUBActionFactory` protocol.
+- The `HUBViewControllerFactory` protocol.
+- The `HUBViewModelLoaderFactory` protocol.
+
+## Layout trait
+
+Layout traits are used to define layout relationships between components. Instead of hard-coding margins and relationships between concrete component implementations - layout traits enables components to work with layout in a more abstract way. Each component simply defines which layout traits that best describe it, in terms of layout, and then an implementation of `HUBComponentLayoutManager` computes the exact margins that should be applied to it.
```
Comment (cerihughes) [inline documentation/Concept guide.md]: Either `Each component simply defines which layout traits best describe it` or `Each component simply defines the layout traits that best describe it`

**Comment by JohnSundell** [comment]
@cerihughes I love nitpicks 🎉 Fixed!

**Pull request merged**